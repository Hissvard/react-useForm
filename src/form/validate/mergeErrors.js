const mergeArray = (fields, form) => {
  if (!fields) return form

  if (!Array.isArray(form)) {
    fields.error = form
    return fields
  }
  if (!Array.isArray(fields)) {
    form.error = fields
    return form
  }

  return form.reduce((acc, value, index) => {
    acc[index] = mergeObject(acc[index], value)

    return acc
  }, fields)
}

const mergeObject = (fields, form) => {
  if (!fields) return form
  if (!form) return fields

  return Object.entries(form).reduce((acc, [key, value]) => {

    if (Array.isArray(acc[key]) || Array.isArray(value)) {
      acc[key] = mergeArray(acc[key], value)
    } else {
      acc[key] = value
    }

    return acc
  }, fields)
}

export const mergeErrors = (fields, form) => {
  return mergeObject(fields, form)
}
