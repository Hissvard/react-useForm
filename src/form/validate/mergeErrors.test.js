import expect from 'expect'
import { mergeErrors } from './mergeErrors'

test('merges with array', () => {
  const formErrors = {
    nickname: 'nickname, error',
    colors: [
      { color: 'color 0 error' },
    ],
  }
  const fieldErrors = {
    colors: [
      null,
      { color: 'color 1 error' },
    ],
  }

  expect(mergeErrors(fieldErrors, formErrors)).toMatchSnapshot()
})

test('missing array on fields ', () => {
  const formErrors = {
    colors: [
      { color: 'color 0 error' },
    ],
  }
  const fieldErrors = {
  }

  expect(mergeErrors(fieldErrors, formErrors)).toMatchSnapshot()
})

test('missing array on form ', () => {
  const formErrors = {
  }
  const fieldErrors = {
    colors: [
      { color: 'color 0 error' },
    ],
  }

  expect(mergeErrors(fieldErrors, formErrors)).toMatchSnapshot()
})

test('more deeply nested arrays data', () => {
  const formErrors = {
    nickname: 'nickname, error',
    colors: [
      { color: 'color 0 error' },
      { shades: 'shades 1 error' },
    ],
  }
  const fieldErrors = {
    colors: [
      null,
      {
        color: 'color 1 error',
        shades: [
          { shade: 'nested shade 1 error' },
        ],
      },
    ],
  }

  const actual = mergeErrors(fieldErrors, formErrors)
  expect(actual).toMatchSnapshot()
  expect(actual.colors[1].shades.error).toMatchSnapshot()
})

test('more deeply nested arrays data opposite', () => {
  const fieldErrors = {
    nickname: 'nickname, error',
    colors: [
      { color: 'color 0 error' },
      { shades: 'shades 1 error' },
    ],
  }
  const formErrors = {
    colors: [
      null,
      {
        color: 'color 1 error',
        shades: [
          { shade: 'nested shade 1 error' },
        ],
      },
    ],
  }

  const actual = mergeErrors(fieldErrors, formErrors)
  expect(actual).toMatchSnapshot()
  expect(actual.colors[1].shades.error).toMatchSnapshot()
})
