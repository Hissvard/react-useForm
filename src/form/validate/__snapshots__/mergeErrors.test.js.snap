// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merges with array 1`] = `
Object {
  "colors": Array [
    Object {
      "color": "color 0 error",
    },
    Object {
      "color": "color 1 error",
    },
  ],
  "nickname": "nickname, error",
}
`;

exports[`missing array on fields  1`] = `
Object {
  "colors": Array [
    Object {
      "color": "color 0 error",
    },
  ],
}
`;

exports[`missing array on form  1`] = `
Object {
  "colors": Array [
    Object {
      "color": "color 0 error",
    },
  ],
}
`;

exports[`more deeply nested arrays data 1`] = `
Object {
  "colors": Array [
    Object {
      "color": "color 0 error",
    },
    Object {
      "color": "color 1 error",
      "shades": Array [
        Object {
          "shade": "nested shade 1 error",
        },
      ],
    },
  ],
  "nickname": "nickname, error",
}
`;

exports[`more deeply nested arrays data 2`] = `"shades 1 error"`;

exports[`more deeply nested arrays data opposite 1`] = `
Object {
  "colors": Array [
    Object {
      "color": "color 0 error",
    },
    Object {
      "color": "color 1 error",
      "shades": Array [
        Object {
          "shade": "nested shade 1 error",
        },
      ],
    },
  ],
  "nickname": "nickname, error",
}
`;

exports[`more deeply nested arrays data opposite 2`] = `"shades 1 error"`;
